Sources/SwiftDocCUtilities/Action/Actions/Convert/ConvertAction.swift

```diff
@@ -198,52 +165,38 @@ public struct ConvertAction: AsyncAction {
-                allowArbitraryCatalogDirectories: allowArbitraryCatalogDirectories
-            )
-        } else {
-            configuration.externalMetadata.isGeneratedBundle = true
-            dataProvider = GeneratedDataProvider(symbolGraphDataLoader: { url in
-                fileManager.contents(atPath: url.path)
-            })
-        }
-        
         if let outOfProcessResolver {
             configuration.externalDocumentationConfiguration.sources[outOfProcessResolver.bundleIdentifier] = outOfProcessResolver
             configuration.externalDocumentationConfiguration.globalSymbolResolver = outOfProcessResolver
         }
         configuration.externalDocumentationConfiguration.dependencyArchives = dependencies
         
-        (context as _DeprecatedConfigurationSetAccess?)?.configuration = configuration
-        
-        self.context = try context ?? DocumentationContext(dataProvider: workspace, diagnosticEngine: engine, configuration: configuration)
-        
-        self.converter = DocumentationConverter(
-            documentationBundleURL: documentationBundleURL,
-            emitDigest: emitDigest,
-            documentationCoverageOptions: documentationCoverageOptions,
-            currentPlatforms: currentPlatforms,
-            workspace: workspace,
-            context: self.context,
-            dataProvider: dataProvider,
-            bundleDiscoveryOptions: bundleDiscoveryOptions,
-            sourceRepository: sourceRepository,
-            diagnosticEngine: self.diagnosticEngine,
-            experimentalModifyCatalogWithGeneratedCuration: experimentalModifyCatalogWithGeneratedCuration
+        let inputProvider = DocumentationContext.InputsProvider(fileManager: fileManager)
+        let (bundle, dataProvider) = try inputProvider.inputsAndDataProvider(
+            startingPoint: documentationBundleURL,
+            allowArbitraryCatalogDirectories: allowArbitraryCatalogDirectories,
+            options: bundleDiscoveryOptions
         )
+        self.configuration = configuration
+        
+        self.bundle = bundle
+        self.dataProvider = dataProvider
     }
     
+    let configuration: DocumentationContext.Configuration
+    private let bundle: DocumentationBundle
+    private let dataProvider: DocumentationBundleDataProvider
+    
     /// A block of extra work that tests perform to affect the time it takes to convert documentation
     var _extraTestWork: (() async -> Void)?
 
     /// Converts each eligible file from the source documentation bundle,
     /// saves the results in the given output alongside the template files.
-    public mutating func perform(logHandle: inout LogHandle) async throws -> ActionResult {
+    public func perform(logHandle: inout LogHandle) async throws -> ActionResult {
+        try await perform(logHandle: &logHandle).0
+    }
+    
+    func perform(logHandle: inout LogHandle) async throws -> (ActionResult, DocumentationContext) {
```

Comment:
The `PreviewAction/convert()` uses it [here](https://github.com/swiftlang/swift-docc/pull/1057/files#diff-c911a2f02efabd902aef8c16be198b9746e2f04e0d86c07472b9e912d29c2ce9R165). This makes the convert action responsible for creating a context from the inputs and pass that context to the preview action. Since the preview action receives the context from the convert action, it never needs to create a context itself and never needs to store a context.